File: java/modules/basic/src/main/java/io/v6d/modules/basic/arrow/Table.java
/** Copyright 2020-2023 Alibaba Group Holding Limited.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.v6d.modules.basic.arrow;

import com.google.common.base.Objects;
import io.v6d.core.client.Context;
import io.v6d.core.client.ds.Object;
import io.v6d.core.client.ds.ObjectFactory;
import io.v6d.core.client.ds.ObjectMeta;
import io.v6d.core.common.util.VineyardException;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import lombok.val;
import org.apache.arrow.vector.VectorSchemaRoot;

public class Table extends Object {
    private final int rows;
    private final int columns;
    private final Schema schema;
    private final List<RecordBatch> batches;
    private List<ObjectMeta> batchMetas;
    private final int batchNum;

    public static void instantiate() {
        RecordBatch.instantiate();
        ObjectFactory.getFactory().register("vineyard::Table", new TableResolver());
    }

    public Table(
            final ObjectMeta meta,
            Schema schema,
            int rows,
            int columns,
            List<RecordBatch> batches) {
        super(meta);
        this.schema = schema;
        this.rows = rows;
        this.columns = columns;
        this.batches = batches;
        this.batchNum = batches.size();
        batchMetas = new java.util.ArrayList<>();
        for (int i = 0; i < this.batchNum; i++) {
            batchMetas.add(meta.getMemberMeta("partitions_-" + i));
        }
    }

    public Schema getSchema() {
        return schema;
    }

    public int getRows() {
        return rows;
    }

    public int getColumns() {
        return columns;
    }

    public List<RecordBatch> getBatches() {
        return batches;
    }

    public int getBatchNum() {
        return batchNum;
    }

    public synchronized RecordBatch getBatch(int index) {
        if (batches.get(index) == null) {
            // Batch is not at local. Find it from remote.
            ObjectMeta batchMeta;
            try {
                batchMeta =
                        Context.getClient()
                                .getMetaData(
                                        meta.getMemberMeta("partitions_-" + index).getId(), true);
            } catch (VineyardException e) {
                Context.println("Get remote batch failed!");
                return null;
            }
            batches.set(index, (RecordBatch) ObjectFactory.getFactory().resolve(batchMeta));
        }
        return batches.get(index);
    }

    public synchronized String[] getHostsOfRecordBatches(int start, int length) {
        if (start + length > batchNum) {
            return null;
        }
        String[] hosts = new String[length];
        for (int i = 0; i < length; i++) {
            hosts[i] = batchMetas.get(start + i).getStringValue("host_");
        }
        return hosts;
    }

    public VectorSchemaRoot getArrowBatch(int index) {
        return batches.get(index).getBatch();
    }

    @Override
    public boolean equals(java.lang.Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Table that = (Table) o;
        if (this.batches.size() != that.batches.size()) {
            return false;
        }
        for (int index = 0; index < this.batches.size(); ++index) {
            if (!Objects.equal(this.batches.get(index), that.batches.get(index))) {
                return false;
            }
        }
        return true;
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(batches);
    }
}

class TableResolver extends ObjectFactory.Resolver {
    @Override
    public Object resolve(final ObjectMeta meta) {
        val ncol = meta.getIntValue("num_columns_");
        val nrow = meta.getIntValue("num_rows_");
        val nbatch = meta.getIntValue("partitions_-size");

        val schema = (Schema) new SchemaResolver().resolve(meta.getMemberMeta("schema_"));

        val batches =
                IntStream.range(0, nbatch)
                        .mapToObj(
                                index -> {
                                    val batch = meta.getMemberMeta("partitions_-" + index);
                                    if (batch.getInstanceId().compareTo(Context.getInstanceID())
                                            == 0) {
                                        return (RecordBatch)
                                                ObjectFactory.getFactory().resolve(batch);
                                    }
                                    return null;
                                })
                        .collect(Collectors.toList());
        return new Table(meta, schema, nrow, ncol, batches);
    }
}

File: go/vineyard/pkg/ds/basic/arrow/table.go
/** Copyright 2020-2023 Alibaba Group Holding Limited.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package arrow

import (
	"fmt"

	"github.com/apache/arrow/go/v11/arrow"
	"github.com/apache/arrow/go/v11/arrow/array"
	"github.com/v6d-io/v6d/go/vineyard/pkg/client"
	"github.com/v6d-io/v6d/go/vineyard/pkg/common/types"
)

type Table struct {
	client.Object
	arrow.Table
}

func (r *Table) Construct(c *client.IPCClient, meta *client.ObjectMeta) (err error) {
	if r.Id, err = meta.GetId(); err != nil {
		return err
	}
	r.Meta = meta
	var schema Schema
	if err = r.Meta.GetMember(c, "schema_", &schema); err != nil {
		return err
	}
	batchNum, err := r.Meta.GetKeyValueUint("partitions_-size")
	if err != nil {
		return err
	}
	batches := make([]arrow.Record, 0, batchNum)
	for i := 0; i < int(batchNum); i++ {
		var batch RecordBatch
		err = meta.GetMember(c, fmt.Sprintf("partitions_-%d", i), &batch)
		if err != nil {
			return err
		}
		batches = append(batches, batch.Record)
	}
	r.Table = array.NewTableFromRecords(schema.Schema, batches)
	return nil
}

type TableBuilder struct {
	Schema        *arrow.Schema
	SchemaBuilder *SchemaBuilder
	NumColumns    uint64
	NumRows       uint64

	Records []arrow.Record
	Batches []types.ObjectID
}

func NewTableBuilder(table arrow.Table) *TableBuilder {
	if table.NumCols() == 0 {
		return NewTableBuilderFromBatches(table.Schema(), []arrow.Record{})
	}
	reader := array.NewTableReader(table, -1)
	batches := make([]arrow.Record, 0)
	for reader.Next() {
		batches = append(batches, reader.Record())
	}
	return NewTableBuilderFromBatches(table.Schema(), batches)
}

func NewTableBuilderFromBatches(schema *arrow.Schema, records []arrow.Record) *TableBuilder {
	return &TableBuilder{
		Schema:  schema,
		Records: records,
	}
}

func (r *TableBuilder) Build(c *client.IPCClient) error {
	r.SchemaBuilder = NewSchemaBuilder(r.Schema)
	if err := r.SchemaBuilder.Build(c); err != nil {
		return err
	}
	r.Batches = make([]types.ObjectID, 0, len(r.Records))
	for _, record := range r.Records {
		r.NumRows += uint64(record.NumRows())
		builder := NewRecordBuilderFromRecord(record)
		if err := builder.Build(c); err != nil {
			return err
		}
		batchId, err := builder.Seal(c)
		if err != nil {
			return err
		}
		r.Batches = append(r.Batches, batchId)
	}
	r.NumColumns = uint64(len(r.Batches))
	return nil
}

func (r *TableBuilder) Seal(c *client.IPCClient) (types.ObjectID, error) {
	if err := r.Build(c); err != nil {
		return types.InvalidObjectID(), err
	}
	meta := client.NewObjectMeta()
	meta.SetTypeName("vineyard::Table")
	schemaId, err := r.SchemaBuilder.Seal(c)
	if err != nil {
		return types.InvalidObjectID(), err
	}
	meta.AddMemberId("schema_", schemaId)
	meta.AddKeyValue("num_columns_", r.NumColumns)
	meta.AddKeyValue("num_rows_", r.NumRows)
	meta.AddKeyValue("batch_num_", uint64(len(r.Batches)))
	meta.AddKeyValue("batch_num_", uint64(len(r.Batches)))
	meta.AddKeyValue("partitions_-size", r.NumColumns)
	for i, batch := range r.Batches {
		meta.AddMemberId(fmt.Sprintf("partitions_-%d", i), batch)
	}
	return c.CreateMetaData(meta)
}
